### ISSUE TITLE
Incorrect treatment of Collections

### ISSUE THREAD

**Issue by jwaataja** [t-rasmud/checker-framework#8]
This `tokenize` method splits a String into tokens. It should return `@PolyDet` array because this method would be deterministic if the input was deterministic and non-deterministic if it was not. However, there are two places where adding to an `ArrayList` says the argument is `@NonDet` when I think it should not be. I was unable to find a simpler case.

This error can be fixed by marking each part of the `ArrayList` as `@NonDet` including the type parameter, and this can be done for the return type as well. However, the method shouldn't always return `@NonDet`, it should depend on the input.

The file appears below:
```
  /**
   * Splits the argument string into an array of tokens (command-line flags and arguments),
   * respecting single and double quotes.
   *
   * <p>This method is only appropriate when the {@code String[]} version of the arguments is not
   * available &mdash; for example, for the {@code premain} method of a Java agent.
   *
   * @param args the command line to be tokenized
   * @return a string array analogous to the argument to {@code main}.
   */
  // TODO: should this throw some exceptions?
  public static String[] tokenize(String args) {

    // Split the args string on whitespace boundaries accounting for quoted
    // strings.
    args = args.trim();
    List<String> argList = new ArrayList<String>();
    String arg = "";
    char activeQuote = 0;
    for (int ii = 0; ii < args.length(); ii++) {
      char ch = args.charAt(ii);
      if ((ch == '\'') || (ch == '"')) {
        arg += ch;
        ii++;
        while ((ii < args.length()) && (args.charAt(ii) != ch)) {
          arg += args.charAt(ii++);
        }
        arg += ch;
      } else if (Character.isWhitespace(ch)) {
        // System.out.printf ("adding argument '%s'%n", arg);
        argList.add(arg);
        arg = "";
        while ((ii < args.length()) && Character.isWhitespace(args.charAt(ii))) {
          ii++;
        }
        if (ii < args.length()) {
          // Encountered a non-whitespace character.
          // Back up to process it on the next loop iteration.
          ii--;
        }
      } else { // must be part of current argument
        arg += ch;
      }
    }
    if (!arg.equals("")) {
      argList.add(arg);
    }

    String[] argsArray = argList.toArray(new String[argList.size()]);
    return argsArray;
  }
```
and the error is:
```
Baz.java:36: error: [argument.type.incompatible] incompatible types in argument.
        argList.add(arg);
                    ^
  found   : @NonDet String
  required: @Det String
Baz.java:51: error: [argument.type.incompatible] incompatible types in argument.
      argList.add(arg);
                  ^
  found   : @NonDet String
  required: @Det String
```

The full error is attached below:
[Baz_errors.txt](https://github.com/t-rasmud/checker-framework/files/2142035/Baz_errors.txt)

**Comment by JasonWaataja** [comment]
I made a much simpler case where this comes up. It is with the following method:
```java
public static List<String> copyList(List<String> strings) {
    List<String> copy = new ArrayList<>();
    for (String s : strings) {
        copy.add(s);
    }
    return copy;
}
```
which gives the error:
```
ListBug.java:7: error: [argument.type.incompatible] incompatible types in argument.
            copy.add(s);
                     ^
  found   : @PolyDet String
  required: @Det String
```

It seems like the local List is not treated correctly.

**Comment by t-rasmud** [comment]
Fixed in 67bde84b49545c94017bfa4a3921a56539088769
